### ISSUE TITLE
node v0.11.* build

### ISSUE THREAD

**Issue by TEugen** [barrysteyn/node-scrypt#39]
Any plans for supporting node v0.11.*?

**Comment by underbluewaters** [comment]
Any news on this? I'm trying to decide whether to use another library.

Sorry I know these messages are annoying, but I have no c++ chops to contribute.

**Comment by tellnes** [comment]
You should reconsider not using the nan project. nan does not solves all issues, but it solves a lot of them. Also, in the io.js world there is talk about shipping nan with io.js (https://github.com/iojs/io.js/issues/725#issuecomment-73068796).

**Comment by barrysteyn** [comment]
Yeah - I might do that. But I am still trying to go for a native node solution, then none of these addon problems will be present.

**Comment by tellnes** [comment]
https://github.com/iojs/io.js/issues/456

**Comment by barrysteyn** [comment]
@tellnes I really like the IO project, and I want to support it. If you can do some investgation or help me, I would be keen to get this going. Does this really only need nan?

**Comment by tellnes** [comment]
I can't answer you if all you need is nan because that is a question about if nan exposes the apis you need to do the binding. But if nan exposes those apis, then that should be all you need.

I think the plan is that if v8 changes its api and it can be fixed for node-scrypt by only using a patched version of the nan module, then io.js will do a minor version bump. If it can not be fixed in nan and you need to patch node-scrypt, then io.js will do a major version bump.

**Comment by barrysteyn** [comment]
Hmmm. Seems reasonable. Tell me something, has io got more excitement at
the moment compared to node?

**Comment by tellnes** [comment]
node.js v0.12 was just released. It is far behind io.js. But as of now both are still api compatible. nan does support both io.js and node.js. I think people still believes in the possibility of a merge in the feature, which would be the best for the community. If that is made possible, then you can look at io.js as a bug pull request to node.js.

io.js is how a lot of people wants node.js development to work. Including most of the core contributors expect for those working for Joyent.

**Comment by tellnes** [comment]
Ref if nan is all you need: If nan does not expose the api you need, than you either open a pull request to nan to make it support it or you use the v8 api directly and hope you does not get a new issue like this one if v8 changes its api.

**Comment by barrysteyn** [comment]
Okay, just for all to know: I am starting to actively work on this issue.

**Comment by BrandonZacharie** [comment]
+1 for 0.12 support

**Comment by dmmalam** [comment]
+1

**Comment by mike-marcacci** [comment]
Just adding my enthusiastic +1 for node v0.12 / io.js 1.2.x support. This is the last package I need updated before I can make the switch in all of my projects :)

**Comment by barrysteyn** [comment]
Its coming @mike-marcacci I need to go over the entire source code. I got about half way over the weekend.

**Comment by mike-marcacci** [comment]
@barrysteyn that is so great to hear! Thanks so much for your weekends

**Comment by mariusk** [comment]
+1 thx

**Comment by barrysteyn** [comment]
Just a quick update for everyone: I have been diving into nan. There are many things I need to use that are not documented, so I am forced to look at the nan source code. Things are coming along quite well: I may finish things by the end of this weekend. If I am not able to finish by the end of this weekend, I should finish by the end of next weekend.

**Comment by barrysteyn** [comment]
Another quick update: It turns out that I underestimated the size of the work needed to accomplish this task (I think that was why I put it off in the first place). Seeing as this is crypto code we are dealing with, I need to be very careful how things are implemented. The good news is that things are coming along. In my last comment, I said I should finish by the end of next weekend, and I still hope to keep that promise, but being honest, I may not be able to do this.

Lets see...

**Comment by mike-marcacci** [comment]
Thanks so much for all your work and time @barrysteyn! If you're ever in Southern California... I owe you a beer. If you think there's a way for others to contribute without stepping on your toes or interfering with your vision, let us know!

**Comment by tellnes** [comment]
+1 what @mike-marcacci said except I live in Trondheim, Norway.

**Comment by barrysteyn** [comment]
Hi @tellnes and @mike-marcacci - Thank you so much for this offer. Honestly, just knowing that there are people who are using my software is enough payment for me.

If this were an ordinary project, I would be very happy to accept help from you guys. The problems is that is is crypto code. If you guys are to help, you must know Colin Percival's C scrypt library inside out and also Node (as well as its buffers and its security flaws). I am lucky in the sense that I live close to Colin (we both live in Vancouver) and so I am able to hit him up with any Questions. I am also in constant contact with Node's core devs to see if what I am doing will be a security flaw.

The part that is taking long is actually the Node bit now - I need to ensure that what I am doing is not going to be a security flaw. Actually, I started the process of writing scrypt from scratch in Node's javascript (so no c++ addon), which would by and large solve lots of these problems. The thing is, I had no idea this library would be so popular. I think it was because I really put lots of effort in it to make sure that it worked and was secure. So I put the Node javascript project on hold, and I am now doing this.

In short: Thanks very much for your offers of help, but I don't think I am able to use them...

**Comment by timshadel** [comment]
@barrysteyn I'm really grateful that you're doing this work as well. I know Colin's C scrypt code pretty well, but not as much the C++ Node part. Honestly, the thing that's enticing about this library is the C++ part. I tried out a pure JS version a couple weeks back, and the performance was terrible in comparison. We're using your library when authenticating a certain set of API calls, and the extra 500ms that library took just won't work for us. Your work is helping us keep better security while writing good apps. :+1:

**Comment by barrysteyn** [comment]
@timshadel Excellent, thanks so much. I promise, the new scrypt module is coming along. I am actively working on it, its just that I am also an engineer lead at a huge multinational software company. So time... But its my top priority now in terms of my open source projects.

**Comment by barrysteyn** [comment]
Okay, as promised, I have something that works for both Node 0.10 and 0.12. But I need people to help me out by just testing it. Can anyone who is interested in giving me feedback please clone the *nan-update* branch and test the module from there. Also, please can you put your feedback on this github issue. Thanks - I am going to sleep now :)

**Comment by tellnes** [comment]
You could start by changing the engines field in package.json for that branch. npm currently fails with the following error:

```
npm ERR! notsup Unsupported
npm ERR! notsup Not compatible with your version of node/npm: scrypt@3.0.1
npm ERR! notsup Required: {"node":">= 0.10 < 0.11"}
npm ERR! notsup Actual:   {"npm":"2.6.1","node":"1.4.2"}
```

**Comment by barrysteyn** [comment]
@tellnes Good catch. Just committed it (a7763071cd3effeb8c0d1a88309a13950b490225) - please can you check again.

**Comment by tellnes** [comment]
Also, `engineStrict` will be removed from npm, https://github.com/npm/npm/commit/7c81a5f5f058941f635a92f22641ea68e79b60db.

**Comment by tellnes** [comment]
It still does not work to run `npm install barrysteyn/node-scrypt#nan-update`, but that could be a caching problem or a bug in npm.

I've cloned the repo and all tests passes on io.js v1.4.2 for me.

**Comment by dpatti** [comment]
I'm getting a segfault with the following repro:

```javascript
var scrypt = require('scrypt');
var scryptParams = scrypt.params(0.1, 0, 33554432);

var hash = new Buffer("c2NyeXB0AA8AAAAIAAAAATxo5F67eRrK887t3wQ3bdnd7HX+N3HlqrhdfCE3fq1KMEx857Qw2RVC9hmPCbfL675hBAbDfLzKdZfBVQSXGI2mFZ6WFtt/EUcYIvielDhn");
var password = new Buffer("080c7323164b6a1ea5a5b9f85db9481ecc61aa415ec1b8b6fbf82f2d222ede92");

scrypt.verify(hash, password);
```

This is node 0.12 on ubuntu x64. If you can't repro, I would be happy to dig into it more.

**Comment by tellnes** [comment]
Running `npm rebuild` from an other package which depends on scrypt fails.

```
cpuid.node seems to not have been built. Run npm install.
gyp: Call to 'node sse-discover.js' returned exit status 0. while trying to load binding.gyp
```

Running `npm rebuild cpuid` first and then `npm rebuild scrypt` after seems to work.

**Comment by barrysteyn** [comment]
@dpatti I cannot reproduce your error. You are going to have to dig some more to help me out. I have almost the identical setup. Your test suite will throw an exception, but it should not segfault.

**Comment by barrysteyn** [comment]
@tellnes I know about that error, but I am going to fix this up in another issue. Right now, I just need to know that the core module works. So if you get that error, can you please just manually install cpuid for now. Also, I am going to remove CPUID completely, instead, I am going to detect on my own for SSE compatibility.

**Comment by tellnes** [comment]
No problem. I'm just reporting it.

**Comment by barrysteyn** [comment]
@tellnes I have removed engineStrict, replaced it with engine-strict. Thanks.

**Comment by tellnes** [comment]
They are removing support for `engineStrict` in package.json and is saying you should use npm config instead. In practice that means the the decision is in the hands of the user and not the module author. See https://github.com/npm/npm/issues/7171#issuecomment-70609581 for details.

**Comment by barrysteyn** [comment]
Hmmm - can npm config get this from package.json? If so, I am going to leave my change be. If not, then I will remove it.

**Comment by tellnes** [comment]
Not that I'm aware of, but I'm not an expert on npm.

**Comment by dpatti** [comment]
I forgot about the encoding on the buffer in my original repro, but even with it (I updated my last comment) I am still getting a segfault on 0.12 but not 0.10. I will admit my C++ debugging chops are not what they used to be, so in my naive approach, I have isolated the problem here:

```c++
  HashInfo(Handle<Object> config) : hash_ptr(NULL), key_ptr(NULL), keySize(0) { 
    std::cout << "HashInfo()\n";
    v8::Local<v8::String> a = NanNew<String>("_hashEncoding");
    std::cout << "1\n";
    v8::Local<v8::Value> b = config->GetHiddenValue(a);
    std::cout << "2\n";
    // The following line causes a segfault:
    v8::Local<v8::Uint32> c = b->ToUint32();
    std::cout << "3\n";
    int d = c->Value();
    std::cout << "value " << d << "\n";
    // Snip...
```

If I run this same code with `_keyEncoding` instead, it works. It looks to me like `CreateScryptConfigObject` [does not initialize `_hashEncoding`][1], but I also have no idea how that would only cause a segfault for me.

[1]: https://github.com/barrysteyn/node-scrypt/blob/d33dff9f9913e7af4bdb1850e5522b09c555ca62/src/node-boilerplate/scrypt_config_object.cc#L133-L136

**Comment by barrysteyn** [comment]
Hi @dpatti 

Thanks for doing the investigation. Short story: I know why it is segfaulting, and I have fixed the problem (get the latest from the nan-update branch).

Long story: In order to hide encoding settings from JS land, I use something called SetHiddenValue (it is a V8 function) that hides properties from users, it can only be accessed by the C++ api. In previous versions of V8, if you accessed a hidden value that was not present, it just defaulted to 0, which turns out to be a valid ENUM value for Node's encoding (I think it is ascii). But the latest version of V8 causes a segfault, and so I need to intialize it. It's acting weird though because I did not initialize the HashEncoding (I forgot about it) yet it was working with me and others.

Anwyways, it's good practice to initialize your things, and this fix should sort you out. Please can you test it for me again.

Thanks

**Comment by dpatti** [comment]
My original repro case is working now, and all of our tests are passing. :+1:

**Comment by dpatti** [comment]
I am running into another issue. We keep modules checked in to source control after installing with `--ignore-scripts`. We always do an `npm rebuild` to compile all dependencies. In the case of the newer scrypt, we have this problem:

```
$ npm install barrysteyn/node-scrypt#nan-update --ignore-scripts
scrypt@3.0.1 node_modules/scrypt
├── nan@1.7.0
└── cpuid@0.1.2

$ npm rebuild

> scrypt@3.0.1 install /home/doug/dev/scrypt-install-test/node_modules/scrypt
> node-gyp rebuild

child_process: customFds option is deprecated, use stdio instead.
cpuid.node seems to not have been built. Run npm install.
gyp: Call to 'node sse-discover.js' returned exit status 0. while trying to load binding.gyp
```

What happens is `cpuid` prints to stderr, which causes gyp to stop and fail regardless of the exit code. And honestly, I don't know how to fix this. I don't know of another module that used a native dependency during its own build process. I'm sure it could be fixed in npm by reversing the order of the rebuild to do dependencies before dependents, but I have no idea how long that would take. Do you have any insights on what could be done to mitigate this within scrypt?

**Comment by gavinhungry** [comment]
Cloned the *nan-update* branch, built successfully and all tests (216) passed with node v0.12.0 (Linux). Thanks for all your work!

**Comment by alemhnan** [comment]
I just tried to use the *nan-update* branch on heroku and it worked without strange problems. Although locally I do have experienced some of the problems mentioned above in this thread. In heroku I explicit included cpuid, not sure if that was needed. 
I tried both node versions (0.10.36 and 0.12.0), locally and on heroku as well.

Keep up the good work!!

**Comment by barrysteyn** [comment]
@dpatti This seems to be a standard problem with cpuid. I am going to change the way the module detected SSE2 availability thereby getting rid of cpuid, but that is not going to happen for this issue.

**Comment by barrysteyn** [comment]
@timshadel Hi Tim - please can you test the latest nan-update branch for me and tell me if the performance is still okay for you?

**Comment by timshadel** [comment]
The performance looks just fine to me. It is slower, but only slightly, so it's well within the timeframes we need to hit.

**Comment by barrysteyn** [comment]
@timshadel I have an idea why it is slower (that's why I asked you to test it). Let me apply some fixes and see if it makes it faster. I will probably move my dev to a new branch for this. I will try finish this week with the fixes

**Comment by timshadel** [comment]
Interestingly, when I use the _nan-update_ branch on node 0.10.35 the performance is essentially equal to the current version of scrypt (0.10 only), but when I run it on 0.12.0, the performance drops slightly. So it may not be scrypt#nan-update at all, but something else in my setup.

**Comment by barrysteyn** [comment]
@timshadel Hmmmm, then I don't think my fix will work. It could be a node 0.12 thing. By how much does the performance drop by?

**Comment by timshadel** [comment]
OK. After isolating the scrypt verification, I collected timing data on scrypt-3.0.1@node-0.10.35, scrypt-nan-update@0.10.35, and scrypt-nan-update@0.12.0. I did 3 runs. For each run I generated 300 hashes, and then looped over them and timed how long each successful verify call took under each environment. I then used a two-sided t-test against the pairs of environments to assess if the timing samples were statistically the same. Most of the time they came out statistically the same. On the times they didn't their raw statistics (medians, avg, range, etc) were very close (< 2ms off). So those times most likely were disrupted by OS/processor/background stuff, in my opinion.

Whatever the slight slowdown was I saw in my tests, it wasn't due to scrypt. Its raw performance seems very stable and usually indistinguishable from the former version.

**Comment by barrysteyn** [comment]
@timshadel Excellent. This is good to hear.

**Comment by barrysteyn** [comment]
This issue is now officially closed in 223f039f410d3f80903ad7f0b0bf71e7956fb8a4

**Comment by thehydroimpulse** [comment]
@barrysteyn Does this bring support for 0.12 only or also io.js?

**Comment by barrysteyn** [comment]
Should work for io.js as well (although I have not tested it). Looking at the above comments, you can see other people have tested it and it works fine. But I aim to support io.js - infact, I am much more keen on it than I am of node. In short: Yes, it will support io

**Comment by taoeffect** [comment]
Hey @barrysteyn (and everyone else helping with this project), I just want to say that I really appreciate the work you guys are doing on this. It's an important project, so thank you for the recent activity, fixes, etc.! :+1:

**Comment by c-geek** [comment]
@barrysteyn All tests passed for me on nan-update too. Thank you very much for your work, you've come along with the right time!

@taoeffect You again!

**Comment by barrysteyn** [comment]
Hi Everyone. Thanks for the support. Knowing that people are using my software makes all the difference. If you want to say thanks, the best way I can think of is to just send me an email (barry.steyn@gmail.com) and tell me how you guys are using my software.

**Comment by dmmalam** [comment]
Thanks for building this.

Dharmesh Malam
